tensorflow/compiler/mlir/xla/transforms/legalize_tf.cc

```diff
@@ -4448,10 +4450,28 @@ class GenericConvertUnsortedSegmentReductionOp : public OpRewritePattern<OpTy> {
     // Broadccast the initial value for reduction. This will become the
     // 'operand' parameter to scatter to for the final scatter op.
     Value init = ConcreteClass::GetInitialValue(data_type.getElementType(),
-                                                op.getLoc(), &rewriter);
-    auto broadcasted_init = rewriter.create<mhlo::BroadcastOp>(
-        op.getLoc(), output_type, init,
-        GetI64ElementsAttr(output_shape, &rewriter));
+                                                loc, &rewriter);
+    auto data_shape = data_type.getShape();
+    llvm::SmallVector<Value, 4> shape_values;
+    shape_values.reserve(output_shape.size());
+    shape_values.push_back(rewriter.create<mlir::ConstantOp>(
+        loc, builder.getI64IntegerAttr(output_shape[0])));
+    for (int i = segment_ids_rank; i < data_shape.size(); ++i) {
+      if (output_shape[i - segment_ids_rank + 1] == ShapedType::kDynamicSize) {
```

Comment:
The `DynamicBroadcastInDimOp` also accepts tensors of index value so you can avoid the casts here.

Reply:
Longer term, an alternative would be to use `shape.shape_of` and `shape.to_extent_tensor` here. Those ops essentially compute the same thing and can lower to the code you have here. 

@joker-eph @silvasean at what point do we want to start depending on shape dialect in the tf legalization?

Reply:
> Longer term, an alternative would be to use `shape.shape_of` and `shape.to_extent_tensor` here. Those ops essentially compute the same thing and can lower to the code you have here.
> 
> @joker-eph @silvasean at what point do we want to start depending on shape dialect in the tf legalization?

actually there are some other questions here after lowered to mhlo dialect, which I have discussed with @silvasean .  After lowering to mhlo dialect, I will compile the IR via IREE pipeline. And after some IREE passes, the IR will be converted as follows:
```
// *** IR Dump After mlir::iree_compiler::Shape::(anonymous namespace)::MaterializeShapeCalculationsPass ***

func @main(%arg0: tensor<?x?xf32> {iree.reflection = {}}, %arg1: !shapex.ranked_shape<[?,?]> {iree.reflection = {}}) -> (tensor<10x?xf32> {iree.reflection = {}}, !shapex.ranked_shape<[10,?]> {iree.reflection = {}}) attributes {iree.module.export, iree.reflection = {f = "I10!B7!d-1d-1R10!B7!d10d-1", fv = "1"}} {
  %0 = mhlo.constant dense<0.000000e+00> : tensor<f32>
  %c10_i64 = constant 10 : i64
  %c1 = constant 1 : index
  %1 = shapex.tie_shape %arg0, %arg1 : tensor<?x?xf32>, !shapex.ranked_shape<[?,?]>
  %2 = dim %1, %c1 : tensor<?x?xf32>
  %3 = index_cast %2 : index to i64
  %4 = tensor_from_elements(%c10_i64, %3) : tensor<2xi64>
  %5 = "shapex.from_extent_tensor"(%4) : (tensor<2xi64>) -> !shapex.ranked_shape<[?,?]>
  %6 = "shapex.ranked_broadcast_in_dim"(%0, %5) {broadcast_dimensions = dense<[]> : tensor<0xi64>} : (tensor<f32>, !shapex.ranked_shape<[?,?]>) -> tensor<10x?xf32>
  %7 = shapex.tie_shape %6, %5 : tensor<10x?xf32>, !shapex.ranked_shape<[?,?]>
  return %7, %5 : tensor<10x?xf32>, !shapex.ranked_shape<[?,?]>
}
```

Notice the IR, this line `return %7, %5 : tensor<10x?xf32>, !shapex.ranked_shape<[?,?]>` must be wrong, because the function return type is `tensor<10x?xf32>, !shapex.ranked_shape<[10,?]>`.  ([?, ?] mis-match [10, ?]). 
The reason is that in IREE, when create `shapex.from_extent_tensor` operation, it will generate shape, and all the dim values are -1(actually is 10 and -1, from the `tensor_from_elements` returned tensor ). In fact, the first value in tensor_from_elements tensor is a constant value `%c10_i64`, but `shapex.from_extent_tensor` can not recognize this.

I don't know whether the mhlo dialect is wrong or this is the bug of IREE.

If it's reasonable to lower `tf.UnsortSegmentSum` to `mhlo.dynamic_braoadcast_in_dim` and `mhlo.scatter`, this means we should modify IREE code .
Or, maybe we should not lower `tf.UnsortSegmentSum` to `mhlo.dynamic_braoadcast_in_dim`  ?

Reply:
The verifier for shapex.tie_shape is too strict. Note that 10x? is compatible with ?x?.

You should be able to fix this by changing .equals [here](https://github.com/google/iree/blob/a33973ebfa725014da83b5905f7dd6a8be4ce3b4/iree/compiler/Dialect/Shape/IR/ShapeOps.cpp#L52) with [mlir::verifyCompatibleShape](https://github.com/llvm/llvm-project/blob/7d076e19e31a2a32e357cbdcf0183f88fe1fb0fb/mlir/include/mlir/IR/TypeUtilities.h#L48)

Reply:
> The verifier for shapex.tie_shape is too strict. Note that 10x? is compatible with ?x?.
> 
> You should be able to fix this by changing .equals [here](https://github.com/google/iree/blob/a33973ebfa725014da83b5905f7dd6a8be4ce3b4/iree/compiler/Dialect/Shape/IR/ShapeOps.cpp#L52) with [mlir::verifyCompatibleShape](https://github.com/llvm/llvm-project/blob/7d076e19e31a2a32e357cbdcf0183f88fe1fb0fb/mlir/include/mlir/IR/TypeUtilities.h#L48)

yeah, I have fix the verifier for shapex.tie_shape, but the returned type of main function are `tensor<10x?xf32>` and  `!shapex.ranked_shape<[10,?]>`. but notice the last line, the returned value are `tensor<10x?xf32>` and  `!shapex.ranked_shape<[?,?]>`. the `!shapex.ranked_shape<[?,?]>` is from the `shapex.from_extent_tensor`.

Reply:
> The `DynamicBroadcastInDimOp` also accepts tensors of index value so you can avoid the casts here.

fixed